File: bsdiff/src/main/java/com/dhy/apkpatch/bsdiff/PatchMaterial.java
package com.dhy.apkpatch.bsdiff;

import java.io.File;
import java.util.Arrays;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PatchMaterial {
    ApkFile newApk;
    ApkFile oldApk;

    public boolean isValid() {
        return newApk.name.equals(oldApk.name) && newApk.branch.equals(oldApk.branch);
    }

    static PatchMaterial getPatchMaterial(Config config) {
        Pattern nameReg = Pattern.compile(config.apkFileNameReg);
        ApkFile newApk = getApk(nameReg, config.oldApkFolder);
        ApkFile oldApk = getApk(nameReg, config.newApkFolder);
        if (newApk == null) throw new IllegalArgumentException("newApk not found");
        if (oldApk == null) throw new IllegalArgumentException("oldApk not found");
        if (newApk.vc < oldApk.vc) {
            ApkFile temp = newApk;
            newApk = oldApk;
            oldApk = temp;
        }
        PatchMaterial patchFile = new PatchMaterial();
        patchFile.newApk = newApk;
        patchFile.oldApk = oldApk;
        return patchFile;
    }

    private static ApkFile getApk(Pattern nameReg, String folder) {
        File file = new File(folder);
        if (file.list() == null) return null;
        return Arrays.stream(Objects.requireNonNull(file.list())).map(name -> {
            Matcher matcher = nameReg.matcher(name);
            if (matcher.find()) {
                ApkFile apkFile = new ApkFile();
                apkFile.name = matcher.group("name");
                apkFile.vc = Integer.parseInt(matcher.group("vc"));
                apkFile.branch = matcher.group("branch");
                apkFile.file = new File(file, name);
                return apkFile;
            } else return null;
        }).filter(Objects::nonNull).findFirst().get();
    }
}

File: bsdiff/src/main/java/com/dhy/apkpatch/bsdiff/PatchMaterial.kt
package com.dhy.apkpatch.bsdiff

import java.io.File
import java.util.regex.Pattern

class PatchMaterial {
    lateinit var newApk: ApkFile
    lateinit var oldApk: ApkFile

    val isValid: Boolean
        get() = newApk.name == oldApk.name && newApk.branch == oldApk.branch

    companion object {
        @JvmStatic
        fun getPatchMaterial(config: Config): PatchMaterial {
            val nameReg = Pattern.compile(config.apkFileNameReg)
            var newApk = getApk(nameReg, config.oldApkFolder)
            var oldApk = getApk(nameReg, config.newApkFolder)
            requireNotNull(newApk) { "newApk not found" }
            requireNotNull(oldApk) { "oldApk not found" }
            if (newApk.vc < oldApk.vc) {
                val temp: ApkFile = newApk
                newApk = oldApk
                oldApk = temp
            }
            val patchFile = PatchMaterial()
            patchFile.newApk = newApk
            patchFile.oldApk = oldApk
            return patchFile
        }

        private fun getApk(nameReg: Pattern, folder: String): ApkFile? {
            val file = File(folder)
            return if (file.list() == null) null
            else file.list()!!.map { name: String ->
                val matcher = nameReg.matcher(name)
                if (matcher.find()) {
                    val apkFile = ApkFile()
                    apkFile.name = matcher.group("name")
                    apkFile.vc = matcher.group("vc").toInt()
                    apkFile.branch = matcher.group("branch")
                    apkFile.file = File(file, name)
                    return@map apkFile
                } else return@map null
            }.filterNotNull().maxBy { it.vc }
        }
    }
}